### PR TITLE
TokenExtensions based position NFT

### PR DESCRIPTION
**Pull request by yugure-orca**
# Benefit
no non-refundable rent & fee
- Mint account is closable and rent will back
- Mint account built-in metadata
- no Metaplex NFT mint fee (0.01 SOL/NFT)

# Remaining design point
Those are not large, but should be determined until the merge.
- ~~keep FreezeAuthority for (near?) future improvement (liquidity lock)~~ --> Yes, position PDA will be FreezeAuthority.
- ~~NFT metadata URI (to change NFT image more Orca2 design friendly)~~ -> We use `https://position-nft.orca.so/meta/<pool address>/<position address>` as metadata URI.
- ~~should we add `getAllPositionAccountsForWallet` now ? (Now we need to find 3 type of positions (position, bundled position, position based TokenExtensions) and it may be confusing for beginners)~~ -> `getAllPositionAccountsByOwner` have been added.

# Design points
## Anchor 0.29 based implementation
Anchor 0.30 provides some useful macro to support TokenExtensions.
https://github.com/coral-xyz/anchor/pull/2789

But we cannot migrate to Anchor 0.30.x due to stack frame size issue with Solana 1.18.

So we have to initialize Mint and TokenAccount(ATA) in `openPositionWithTokenExtensions` manually.
To initialize them manually, we need to receive those account as `UncheckedAccount`.

note: of course this implementation should work with newer Anchor versions.

## NFT metadata
name: OWP xxxx...yyyy  (xxxx and yyyy are first and last 4 letters of mint)
symbol: OWP
uri: `https://position-nft.orca.so/meta/<pool address>/<position address>`

### name
It is difficult to identify a specific Position NFT when the name of NFT is "Orca Whirlpool Position".
This is why we use "OWP xxxx...yyyy" as name.

### uri
To provide more helpful NFT image generated dynamically, we use URI containing both pool address and position address.
note: Its length must be less than 128 bytes to display NFTs in phantom.

## no special instruction for `WithMetadata`
There are `openPosition` and `openPositionWithMetadata`, but we will introduce `openPositionWithTokenExtensions` only.

- There are no large differences in provided accounts. So now we can switch behaviour based on `with_token_metadata` parameter.
- Even if a position Mint doesn't have TokenMetadata extension, it will have MintCloseAuthority extension.

## shared increase/decrease_liquidity, collect_fees, collect_reward, update_fees_and_rewards instructions
For operations to positions other than open and close, conventional instructions (V1 and V2) can be used.
Since we do not need mint in those instructions, we only need to allow Token-2022 based TokenAccounts.

## PDA seeds for TokenExtensions based positions
We use same seeds for TokenExtensions based positions.

```
seeds = [b"position".as_ref(), position_mint.key().as_ref()],
```

So we can use `PDAUtil.getPosition` for both Mint owned by TokenProgram and Mint owned by Token-2022 program.
Down side is that to determine a position is TokenExtensions based, we need to fetch Mint and check its owner program.

In any case, any flow that obtains a Token-2022 token and discovers a Position necessarily makes these distinctions.

If gPA is used to discover a Position, it would be useful to be able to determine the type of Position from the address, but this is not a critical issue, IMO.

## MintCloseAuthority
The authority for MintCloseAuthority extension is Position account (PDA).
So `closePositionWithTokenExtensions` instruction can close Mint account with `invoke_signed` with Position account authority.

Also Position account is used as authority during initial NFT mint process.
This is same to PositionBundle initialization process, and mint authority will be removed at the end of instruction.

## FreezeAuthority
Position account (PDA) will be FreezeAuthority.
At the moment, FreezeAuthority is not used at all.
But we strongly expect that it is used to implement liquidity locking.

## Collection of NFT
It was confirmed that the GroupMemberPointer and GroupMember extensions are required to display NFTs as a group on the wallet. (I asked Phantom).
https://docs.phantom.app/developer-powertools/solana-token-extensions-token22

The team discussed this, but decided that using these extensions would be overkill and decided not to implement the Collection feature.

For Metaplex TokenMetadata, this could be implemented with a `collection` field, but for TokenExtensions it would be a bit more complicated.

### no specific method for `openPositionWithTokenExtensions` on `Whirlpool` class
We added `withTokenExtensions` optional parameter to both `Whirlpool.openPosition` and `Whirlpool.openPositionWithMetadata`.
When `withTokenExtensions` is true, those functions will use `openPositionWithTokenExtensions`.

The default value of this parameter is `false`, but we can change it to `true` with careful announcement because we strongly recommend to use `openPositionWithTokenExtensions` in the future major release (>= 0.14.0).

### CODE REVIEW

File: programs/whirlpool/src/manager/swap_manager.rs
Comment (yugure-orca) [inline programs/whirlpool/src/manager/swap_manager.rs:1]: only changed by `cargo fmt`.

**Review by wjthieme — COMMENTED** [review]
> name: OWP xxxx...yyyy (xxxx and yyyy are first and last 4 letters of mint)

I understand what this is doing but I have 2 concerns:
* If the name is not in the metadata json (which it is not because that one is static) wallets might not display it.
* Mint address potentially does not add any new value since wallets already display this?

> keep FreezeAuthority for (near?) future improvement (liquidity lock)

We might need to add this to all (or none) of the instructions otherwise we'll potentially create a scenario where certain positions are 'lockable' but others are not.

> collection

We could create a collection and add all position under that collection. Most wallets will then display it as a folder/group which might be a nice feature.

**Comment by yugure-orca** [comment]
@wjthieme 
Thanks for the review!

## Collection
As we discussed internally, "collection" feature will not be implemented.
I added the summary of discussion at the PR comment.

I very much agree that it is useful to have them all displayed grouped, although we would not implement it to avoid complexity. Thanks for the proposal!

## OWP xxxx...yyyy
> Mint address potentially does not add any new value since wallets already display this?

hmm, is it possible to transfer a specific position NFT to another wallet ?

I sometimes tried to transfer some test Positions to another wallet, but it was too difficult to identify which position is what I want to transfer.

The following is the screenshot of collection page of Phantom holding several Position NFTs. 🤦‍
screenshot 2024-10-02 13 14 55

Even when I click one of them, Phantom doesn't display mint address, so I need to view it on Solscan. (And popup will be closed and NFT display order may be changed...)

screenshot 2024-10-02 13 15 23

So I think "mint address" (or "position address") will help users to identify a specific Position NFT.

> If the name is not in the metadata json (which it is not because that one is static) wallets might not display it.

I created the following two NFTs based with TokenMetadata extension.
And it worked as intended because wallet app prioritize on-chain metadata. 👍️

```
mint1: 7e5nrZhrSKTUeCKMfTZiWLoL17TUPNR1QLb7cdYkKpoD

name: "TYO 7e5n...KpoD"
symbol: "TYO"
uri: https://qsaqzautc2yrdgecv4lrsttqxmrqd3xqjk36dlpwpnxlt6aafriq.arweave.net/hIEMgpMWsRGYgq8XGU5wuyMB7vBKt-Gt9ntuufgALFE

mint2: Ns2u2rRV7E6N7WwgdeJaPi4gRShnAB6oGTviL9iWc9h

name: "TYO Ns2u...Wc9h"
symbol: "TYO"
uri: https://qsaqzautc2yrdgecv4lrsttqxmrqd3xqjk36dlpwpnxlt6aafriq.arweave.net/hIEMgpMWsRGYgq8XGU5wuyMB7vBKt-Gt9ntuufgALFE
```

In metadata JSON file, name is "Test Token TOKYO". But Phantom and Backpack displayed "TYO xxxx...yyyy" in their collection page.
Phantom clearly say they prioritize on-chain data if name and symbol field are provided from both on-chain and off-chain.

https://docs.phantom.app/best-practices/tokens/collectibles-nfts-and-semi-fungibles#displaying-an-individual-collectible


> Displaying an Individual Collectible
> When displaying the detail view of an individual collectible, Phantom will prioritize on-chain data in the Metadata Account over off-chain JSON linked via the uri field. This impacts both the name and symbol field which appears in both locations.

Phantom
screenshot 2024-10-02 12 57 14

Backpack
screenshot 2024-10-02 13 00 21

Solflare
cannot see NFTs based on TokenExtensions.

File: legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts [hunk 1]
```diff
@@ -355,20 +363,35 @@ export class WhirlpoolImpl implements Whirlpool {
...
-      positionPda,
-      metadataPda,
-      positionMintAddress: positionMintPubkey,
-      positionTokenAccount: positionTokenAccountAddress,
-      whirlpool: this.address,
-      tickLowerIndex: tickLower,
-      tickUpperIndex: tickUpper,
-    });
-    txBuilder.addInstruction(positionIx);
+    if (withTokenExtensions) {
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts:366]: Very helpful! 🙏
How does `useTokenExtensions` parameter sound to you ?

**Comment by wjthieme** [comment]
> And it worked as intended because wallet app prioritize on-chain metadata

Cool! That was what I was potentially concerned about. Did you also consider adding the position/mint address as attributes to the metadata but IIRC has some downsides:
* 1 added instruction for each attribute
* Increased rent on the mint (but it is refundable on close so not the worst)

File: programs/whirlpool/src/util/token_2022.rs [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        &[
+            position_mint.to_account_info(),
+            authority.to_account_info(),
+            token_2022_program.to_account_info(),
+        ],
+    )?;
+
+    if with_token_metadata_extension {
+        // initialize MetadataPointer extension
+        // authority: None
```
Comment (philcchen) [inline programs/whirlpool/src/util/token_2022.rs:79]: if we are setting the UpdateAuthority on the TokenMetadata anyways, is there any downside to also setting the authority here?

File: Cargo.lock [hunk 3]
```diff
@@ -2412,9 +2412,9 @@ dependencies = [
 
 [[package]]
 name = "spl-token-metadata-interface"
-version = "0.2.1"
+version = "0.2.0"
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply Cargo.lock:2415]: I believe that this has been imported indirectly, but that the feature has not been used.

Now, we need to add `spl-token-metadata-interface` as one of direct dependency.
But `0.2.x (x > 0)` have been yanked and  `>=0.3` requires Solana 1.18.
So 0.2.0 is the only option.

- https://crates.io/crates/spl-token-metadata-interface/0.2.1
- https://crates.io/crates/spl-token-metadata-interface/versions

This yank that prevents Solana 2.0 from being selected is very confusing.🤦‍

File: programs/whirlpool/src/util/token_2022.rs (hunk 2 above)
Comment (yugure-orca) [reply programs/whirlpool/src/util/token_2022.rs:79]: I don't think there is any clear downside.
I think `None` is "a bit" safer than `Some(auth)` here.

Do you think setting WHIRLPOOL_NFT_UPDATE_AUTH would avoid our future regret ?

background:
I thought that there is no possibility of modification, and so I set Authority to None.
The case where MetadataPointer refers to something other than Mint itself would be if we stop using the TokenMetadata extension and start using Metaplex TokenMetadata again, which I thought was unlikely to happen.

As for the metadata, we have made changes to add “DO NO BURN” in the past. So we need to be able to change it.

File: programs/whirlpool/src/util/token_2022.rs [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+                ExtensionType::MetadataPointer,
+            ]
+        } else {
+            &[ExtensionType::MintCloseAuthority]
+        },
+    )?;
+
+    let lamports = Rent::get()?.minimum_balance(if with_token_metadata_extension {
+        // we need to add rent for TokenMetadata extension to reallocate space
+        space + WP_2022_METADATA_MAX_LEN
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/token_2022.rs:38]: good point!
I will try measuring metadata and using exact rent.

We can extend Metadata space if needed by sending SOL and update metadata (including realloc).
https://solscan.io/tx/3AHBSNDa4f5Gdxswn4eqA918vkxEDUq4aNBMnfqwoFrhvtU6Tpz9H8Xv6F7h978u5KCFQrXiW6iDBM4Avo8tg9Q9

File: legacy-sdk/whirlpool/src/instructions/close-position-with-token-extensions-ix.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ * @category Instructions
+ * @param context - Context object containing services required to generate the instruction
+ * @param params - ClosePositionWithTokenExtensionsParams object
+ * @returns - Instruction to perform the action.
+ */
+export function closePositionWithTokenExtensionsIx(
+  program: Program<Whirlpool>,
+  params: ClosePositionWithTokenExtensionsParams,
+): Instruction {
+  const {
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply legacy-sdk/whirlpool/src/instructions/close-position-with-token-extensions-ix.ts:38]: Good catch! 🙏

File: legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts [hunk 6]
```diff
@@ -355,20 +363,35 @@ export class WhirlpoolImpl implements Whirlpool {
...
+      const positionIx = openPositionWithTokenExtensionsIx(this.ctx.program, {
```
[the remark this answers is not in the document]
Comment (yugure-orca) [reply legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts:367]: yep, it makes sense. 👍️

[the remark this answers is not in the document]
Comment (yugure-orca) [reply legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts]: updated

File: legacy-sdk/whirlpool/src/instructions/close-position-with-token-extensions-ix.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (yugure-orca) [reply legacy-sdk/whirlpool/src/instructions/close-position-with-token-extensions-ix.ts]: updated

File: programs/whirlpool/src/util/token_2022.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yugure-orca) [reply programs/whirlpool/src/util/token_2022.rs]: implemented at commit: https://github.com/orca-so/whirlpools/pull/341/commits/52b3a4f4661bb33643ff74c559edece401439a48 
Now rent is minimum (necessary and sufficint) . 🫡

File: legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (yugure-orca) [reply legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts]: Finally, `tokenProgramId` was adopted.
https://github.com/orca-so/whirlpools/pull/341/commits/5fbe6813a53fba5b92d93cb83ca13d558be74b04

**Review by wjthieme — APPROVED** [review]
🚢

**Pull request merged**